### PR TITLE
stdlib: support std_concat()

### PR DESCRIPTION
**Pull request by Rtoax**
Support stdlib `str_concat()`, see PR: [stdlib: support get current working directory](https://github.com/bpftrace/bpftrace/pull/5231)

This PR contains two separate commits (non-squash merge is better):

1. Assembles `bpffeature` into `cflags` and passes it to the stdlib C compiler. This is to allow stdlib C to determine if `bpf-helper` is available;

2. Adds the `bpf_concat()` function to concatenate a string and return a new string; also, if `bpf_loop()` is supported, stdlib C's `__bpf_strnlen()` will replace the `for` loop with `bpf_loop`.

Usage:

        $new = str_concat("hello", " world");
        $new = str_concat($new, "!!!");
        $new = str_concat("INFO: ", $new);
    
        printf("%s\n", $new);
    
        INFO: hello world!!!

### CODE REVIEW

**Comment by Rtoax** [comment]
The force commit above: Add runtime test for `str_concat()` and rebase to master to remove duplicate helpers macro from bpffeature-cflags.

**Review by jordalgo — COMMENTED** [review]
Still reviewing but just as a note: we can get rid of the whole feature check on `bpf_loop`. It was added in kernel 5.17 but the min kernel version is 6.1:
https://github.com/bpftrace/bpftrace/blob/master/docs/dependency_support.md#linux-kernel

File: src/stdlib/strings/strings.bpf.c [hunk 1]
```diff
@@ -12,11 +12,49 @@
...
+};
+
+static int strnlen_cb(__u32 index, void *data)
+{
+  struct strnlen_ctx *ctx = data;
+  if (index > ctx->sz) {
+    return 1;
+  }
+  char ch;
+  // linux >= v5.5
```
Comment (viktormalik) [inline src/stdlib/strings/strings.bpf.c]: No need for this comment, `bpf_probe_read_kernel` is used through the entire codebase so it must be present.

File: src/stdlib/strings/strings.bpf.c [hunk 2]
```diff
@@ -12,11 +12,49 @@
 
 extern int bpf_strnlen(const char *s__ign, size_t count) __ksym __weak;
 
+#ifdef has_bpf_loop
```
Comment (viktormalik) [inline src/stdlib/strings/strings.bpf.c]: `bpf_loop` can be assumed, we support kernel versions >= 6.1 (see [Dependency support policy](https://github.com/bpftrace/bpftrace/blob/master/docs/dependency_support.md) for details).

OTOH, `bpf_loop` is kind of legacy and `bpf_for` should be preferred. However, it is only available since kernel 6.4 so that would be out of our minimal dependency support. @jordalgo WDYT, should we go with `bpf_loop` and modernize to `bpf_for` later?

File: src/stdlib/strings.bt [hunk 3]
```diff
@@ -127,6 +127,42 @@ macro strstr(haystack, needle)
...
+macro str_concat(s1, $s2) {
+  let $s1 = s1;
+  str_concat($s1, $s2)
+}
+
+macro str_concat(s1, s2) {
+  let $s1 = s1;
+  let $s2 = s2;
+  str_concat($s1, $s2)
+}
```
Comment (viktormalik) [inline src/stdlib/strings.bt:163]: Isn't the last variant sufficient?

File: src/stdlib/strings/strings.bpf.c (hunk 2 above)
Comment (jordalgo) [reply src/stdlib/strings/strings.bpf.c]: > should we go with bpf_loop and modernize to bpf_for later

I think we can do that in a follow-up. We probably will need to re-evaluate our min kernel version soon anyway. For now let's just remove the feature check and assume `bpf_loop` is available.

Comment (viktormalik) [reply src/stdlib/strings/strings.bpf.c]: The thing is that `bpf_for` has a much nicer structure.

Compare
```
struct strnlen_ctx
{
    ...
};

static int strnlen_cb(__u32 index, void *data)
{
    ... actual loop implementation ...
}

long __bpf_strnlen(const char *ptr, size_t max_size)
{
    ...
    struct strlen_ctx ctx = {
        ...
    };
    bpf_loop(max_size, strnlen_cb, &ctx, 0);
    ...
}
```
versus
```
long __bpf_strnlen(const char *ptr, size_t max_size)
{
    ...
    bpf_for(0, max_size) {
        ... actual loop implementation ...
    }
    ...
}
```
If we want to consider bumping minimal supported kernel version past 6.4, this may be a good time to do so because it will simplify the code a lot.

File: src/stdlib/strings.bt (hunk 3 above)
Comment (Rtoax) [reply src/stdlib/strings.bt:163]: It's true that every variant has its reason to exist. However, I reworked the implementation of the last variant (keeping the code concise, though).

File: src/stdlib/strings/strings.bpf.c (hunk 2 above)
Comment (Rtoax) [reply src/stdlib/strings/strings.bpf.c]: I added a TODO for bpf_for in the code, but for now I'm still using bpf_loop()

Comment (jordalgo) [reply src/stdlib/strings/strings.bpf.c]: @viktormalik I don't feel strongly but our current policy is the last 4 LTS kernels which includes 6.1.177

File: src/stdlib/strings/strings.bpf.c [hunk 4]
```diff
@@ -12,19 +12,45 @@
...
+struct strnlen_ctx {
+  const char *str;
+  __u32 sz;
+  __u32 len;
+};
+
+static int strnlen_cb(__u32 index, void *data)
+{
+  struct strnlen_ctx *ctx = data;
+  if (index > ctx->sz) {
```
Comment (viktormalik) [inline src/stdlib/strings/strings.bpf.c]:
```suggestion
  if (index >= ctx->sz) {
```

File: src/stdlib/strings/strings.bpf.c [hunk 5]
```diff
@@ -12,19 +12,45 @@
...
+    return 1;
+  }
+  char ch;
+  bpf_probe_read_kernel(&ch, sizeof(char), (void *)(ctx->str + index));
+  if (ch == '\0') {
+    // terminate the bpf_loop()
+    return 1;
+  }
+  ctx->len++;
```
Comment (viktormalik) [inline src/stdlib/strings/strings.bpf.c]: `bpf_loop` should return the number of iterations executed so we don't need `ctx->len`.

File: tests/runtime/stdlib [hunk 6]
```diff
@@ -71,6 +71,10 @@ NAME strlen
 PROG begin { $a = "hello"; print(strlen($a)); }
 EXPECT 5
 
+NAME str_concat
```
Comment (viktormalik) [inline tests/runtime/stdlib]: Should this rather go to selftests?

Comment (viktormalik) [reply tests/runtime/stdlib]: I meant `tests/self`, those should be significantly faster than the runtime tests.

**Comment by Rtoax** [comment]
CI error on alphine ` Linux version 6.14.4 (root@18ff2c5a2c33) (gcc (Alpine 12.2.1_git20220924-r10) 12.2.1 20220924, GNU ld (GNU Binutils) 2.40) #1 SMP PREEMPT_DYNAMIC Tue Apr 29 19:02:31 UTC 2025`:

```log
....
  [ RUN      ] macro.user macros don't override indirect stdlib macros
  [  FAILED  ] macro.user macros don't override indirect stdlib macros (2454 ms)
  	Command: /home/runner/work/bpftrace/bpftrace/build-ci/tests/../src/bpftrace  --verify-llvm-ir -e 'macro assert_str(exp) { fail("boom") } begin { print(strlen("hi")); }'
  	Expected: 2
  	Found:
  Attached 1 probe\n__BPFTRACE_NOTIFY_PROBES_ATTACHED\n3\n\n\n
  [ RUN      ] macro.user macros don't override stdlib if different signature
  [       OK ] macro.user macros don't override stdlib if different signature (2322 ms)
  [ RUN      ] macro.user macros don't poison stdlib builtin calls
  [  FAILED  ] macro.user macros don't poison stdlib builtin calls (2100 ms)
  	Command: /home/runner/work/bpftrace/bpftrace/build-ci/tests/../src/bpftrace  --verify-llvm-ir -e 'macro fail(x) { x } begin { print(strlen("hi")); }'
  	Expected: 2
  	Found:
  Attached 1 probe\n__BPFTRACE_NOTIFY_PROBES_ATTACHED\n3\n\n\n

....

  [  FAILED  ] stdlib.strlen (2578 ms)
  	Command: /home/runner/work/bpftrace/bpftrace/build-ci/tests/../src/bpftrace  --verify-llvm-ir -e 'begin { $a = "hello"; print(strlen($a)); }'
  	Expected: 5
  	Found:
  Attached 1 probe\n__BPFTRACE_NOTIFY_PROBES_ATTACHED\n6\n\n\n
  [----------] 14 tests from stdlib

....

  [  FAILED  ] 3 tests, listed below:
  [  FAILED  ] macro.user macros don't override indirect stdlib macros
  [  FAILED  ] macro.user macros don't poison stdlib builtin calls
  [  FAILED  ] stdlib.strlen
```

Don't known why, debugging....

**Comment by Rtoax** [comment]
> CI error on alphine ` Linux version 6.14.4 (root@18ff2c5a2c33) (gcc (Alpine 12.2.1_git20220924-r10) 12.2.1 20220924, GNU ld (GNU Binutils) 2.40) #1 SMP PREEMPT_DYNAMIC Tue Apr 29 19:02:31 UTC 2025`:
> 
> ```
> ....
>   [ RUN      ] macro.user macros don't override indirect stdlib macros
>   [  FAILED  ] macro.user macros don't override indirect stdlib macros (2454 ms)
>   	Command: /home/runner/work/bpftrace/bpftrace/build-ci/tests/../src/bpftrace  --verify-llvm-ir -e 'macro assert_str(exp) { fail("boom") } begin { print(strlen("hi")); }'
>   	Expected: 2
>   	Found:
>   Attached 1 probe\n__BPFTRACE_NOTIFY_PROBES_ATTACHED\n3\n\n\n

I figure it out, string length does not include the null terminator '\0'. 

@viktormalik @jordalgo Please review, thanks ;)

File: src/stdlib/strings/strings.bpf.c [hunk 7]
```diff
@@ -12,19 +12,46 @@
...
+    .str = ptr,
+    .sz = max_size,
+  };
+  int len = bpf_loop(max_size, strnlen_cb, &ctx, 0);
+  if (len <= 0) {
+    return len;
   }
-  return sz;
+  // The string length does not include the null terminator '\0'.
+  return len - 1;
```
Comment (viktormalik) [inline src/stdlib/strings/strings.bpf.c]: We also need to catch the case when `len == max_size`:
```suggestion
  return len == max_size ? len : len - 1;
```

Comment (Rtoax) [reply src/stdlib/strings/strings.bpf.c]: Thanks

**Review by viktormalik — APPROVED** [review]
Nice feature, thanks!

File: src/stdlib/strings/strings.bpf.c (hunk 7 above)
Comment (Rtoax) [reply src/stdlib/strings/strings.bpf.c]: @viktormalik Sorry, i found another issue, if `strlen("hi")`, `max_size==3`, then, `strlen("hi")` will return 3. see:

```log
  [  FAILED  ] macro.user macros don't override indirect stdlib macros (2533 ms)
  	Command: /home/runner/work/bpftrace/bpftrace/build-ci/tests/../src/bpftrace  --verify-llvm-ir -e 'macro assert_str(exp) { fail("boom") } begin { print(strlen("hi")); }'
  	Expected: 2
  	Found:
  Attached 1 probe\n__BPFTRACE_NOTIFY_PROBES_ATTACHED\n3\n\n\n
  [ RUN      ] macro.user macros don't override stdlib if different signature
  [       OK ] macro.user macros don't override stdlib if different signature (2290 ms)
  [ RUN      ] macro.user macros don't poison stdlib builtin calls
  [  FAILED  ] macro.user macros don't poison stdlib builtin calls (2337 ms)
  	Command: /home/runner/work/bpftrace/bpftrace/build-ci/tests/../src/bpftrace  --verify-llvm-ir -e 'macro fail(x) { x } begin { print(strlen("hi")); }'
  	Expected: 2
  	Found:
  Attached 1 probe\n__BPFTRACE_NOTIFY_PROBES_ATTACHED\n3\n\n\n

...
  [  FAILED  ] stdlib.strlen (2582 ms)
  	Command: /home/runner/work/bpftrace/bpftrace/build-ci/tests/../src/bpftrace  --verify-llvm-ir -e 'begin { $a = "hello"; print(strlen($a)); }'
  	Expected: 5
  	Found:
  Attached 1 probe\n__BPFTRACE_NOTIFY_PROBES_ATTACHED\n6\n\n\n
```

Therefore, it is necessary to check if the last character read is '\0'. please review ;)

**Comment by Rtoax** [comment]
@viktormalik Please review https://github.com/bpftrace/bpftrace/compare/d8349e4b770bfa398e07a89e96adf70fd8277992..68ca5333b839fb8176c8487eeb798f7ad915e752 for the above `Unresolved conversation`, thanks.

File: src/stdlib/strings/strings.bpf.c [hunk 8]
```diff
@@ -12,19 +12,46 @@
...
+    .str = ptr,
+    .sz = max_size,
+  };
+  int len = bpf_loop(max_size, strnlen_cb, &ctx, 0);
+  if (len <= 0) {
+    return len;
   }
-  return sz;
+  // The string length does not include the null terminator '\0'.
+  return (len == max_size && ctx.ch != '\0') ? len : len - 1;
```
Comment (viktormalik) [inline src/stdlib/strings/strings.bpf.c]: We don't need to store `ctx.ch`, just do
```suggestion
  return (len == max_size && ptr[len - 1] != '\0') ? len : len - 1;
```

**Review by viktormalik — APPROVED** [review]

**Comment by viktormalik** [comment]
@Rtoax ready for merge?
